### PR TITLE
feat: encode `let_fun` using a `letFun` function

### PR DESCRIPTION
**Pull request by kmill**
Switches from encoding `let_fun` using an annotated `(fun x : t => b) v` expression to a function application `letFun v (fun x : t => b)`.

### CODE REVIEW

**Comment by kmill** [comment]
@leodemoura I carried out your suggestion to re-encode `let_fun`, and hopefully it came out in the way you anticipated!

**Comment by leodemoura** [comment]
> I'm not sure if you were expecting it to be for dependent or non-dependent lets.

It is for non-dependent lets.

> I investigated making it be only for non-dependent lets, but I found that the first example in tests/lean/letFun.lean

I can encode all `let_fun`s in this test file using the auxiliary `letFun` function.  Could you please clarify the issue you are having?

```lean
def letFun {α : Sort u} {β : α → Sort v} (v : α) (f : (x : α) → β x) : β v := f v

#check
  letFun (fun x => x * 2) fun f =>
  letFun 1 fun x =>
  letFun (x + 1) fun y =>
  f (y + x)

example (a b : Nat) (h1 : a = 0) (h2 : b = 0) : (letFun (a + 1) fun x => x + x) > b := by
  simp (config := { beta := false }) [h1]
  trace_state
  simp (config := { decide := true }) [h2]
```

**Comment by kmill** [comment]
@leodemoura I've used the wrong terminology -- using `SimpLetCase` for clarification, I mean to say I wasn't sure if you were expecting the `nondepDepVar` or `nondep` encoding.  I got the `nondepDepVar` encoding of `letFun` working in this PR, and it does exactly what you wrote in your code block.

**Comment by kmill** [comment]
@leodemoura I don't have any obstruction here, and I can handle non-constant type families. I just wanted to be sure that you weren't expecting this simplified `letFun` encoding [that you had proposed in your original message](https://github.com/leanprover/lean4/pull/2864#issuecomment-1807176233) since I instead used [this "`nonDepDepVar`" encoding](https://github.com/leanprover/lean4/pull/2973/files#diff-52ef0c67eea613acf6c0b6284063fd7112cdd40750de38365c214abaef246db4R71), which I switched to at the first sign of trouble. The latter was more obvious how to implement. In particular [`mkLetFun`](https://github.com/leanprover/lean4/pull/2973/files#diff-042e743a1251bf5b246b4ff2fcf7c2258550eaf715fbe888baaea034ecedc167R29-R36), to produce the simple encoding, would need to force `β` to not depend on the fvar `x`, even in the presence of metavariables.

Examples like this one definitely work with this PR's implementation:
```lean
#check let_fun n := 5
  (⟨[], Nat.zero_le n⟩ : { as : List Bool // as.length ≤ n })
```
I haven't evaluated whether these sorts of examples appear in the wild, but I would not be surprised if they do.

You are correct that it does not generalize the value in the expected type, so your examples with `{ as : List Bool // as.length ≤ 5 }` do not work. I personally believe generalization would be a surprising feature for `let_fun`, and if that is at some point a requested feature, that it should be either an option (like `let_fun (generalizing := true) n := 5; ...`) or be a new binding syntax.

**Comment by leodemoura** [comment]
@kmill Let's move forward and merge.

**Comment by kmill** [comment]
@leodemoura Sounds good. Let me know if you need me to do anything. (Maybe I should rebase this onto the newest nightly and regenerate the stage0?)

**Comment by semorrison** [comment]
> (Maybe I should rebase this onto the newest nightly and regenerate the stage0?)

Yes please.

**Comment by kmill** [comment]
@semorrison Done.

By the way, this will take a small change to std. I don't have push access to that project, so didn't update the testing branch. Instead, I made my own branch and modified the mathlib testing branch to use that. It was an easy fix: https://github.com/kmill/std4/commit/7ca1c2c8232a72a10ba14e69869d1efc6666b043

File: src/Init/Prelude.lean
Comment (Kha) [inline src/Init/Prelude.lean:74]: This encoding means that even with `config.zeta` false, we reduce `let_fun`s when at default reducibility, correct? Is that intentional or should it be `[irreducible]`?

Comment (kmill) [reply src/Init/Prelude.lean]: Good point, that is something I did not think enough about. I made it be irreducible and added more tests. (I also extended the optimization for ignoring unnecessary lets in `isDefEqQuick` to also consider `let_fun`.)

File: src/Init/Prelude.lean
Comment (Kha) [inline src/Init/Prelude.lean:75]:
```suggestion
Both WHNF and `simp` are aware of `letFun` and can reduce it when zeta reduction is enabled,
```

**Review by Kha — APPROVED** [review]

File: src/Init/Prelude.lean
Comment (Kha) [reply src/Init/Prelude.lean]: @leodemoura I'll wait for your final confirmation on this point before merging

Comment (leodemoura) [reply src/Init/Prelude.lean]: Agreed. We don't want to reduce `let_fun`s when `config.zeta = false` and using the default reducibility.

Comment (leodemoura) [reply src/Init/Prelude.lean]: Agreed. We don't want to reduce `let_fun`s when `config.zeta = false` and using the default reducibility.

**Comment by semorrison** [comment]
@Kha, @kmill, I don't think we should have merged this while it had a `breaks-mathlib` label!

It has indeed broken both Std and Mathlib's `nightly-testing` branches, and without understanding this PR in detail I don't immediately know how to fix them.

(The [discussion on zulip](https://leanprover.zulipchat.com/#narrow/stream/345428-mathlib-reviewers/topic/CI.20failure.20on.20the.20nightly-testing.20branch/near/409042357) might be a better place for follow-up than here.)

**Comment by kmill** [comment]
@semorrison Did you see my message earlier, https://github.com/leanprover/lean4/pull/2973#issuecomment-1853065287 ?

I'll follow up on zulip too.

**Pull request merged**
